interpreter allows declaring methods on non-local types when Go does not

With the following program:

```go
package main

import "fmt"

type Foo = int

func (f Foo) Bar() int { return int(f * f) }

func main() {
	x := Foo(1)
	fmt.Println(x.Bar())
}
```

Go allows methods on types declared as `type Foo int`, but with the `=` it considers `Foo` to be 'non-local' and does not allow methods to be defined.

Expected result:
```console
$ go run ./test.go 
# command-line-arguments
./test.go:7:6: cannot define new methods on non-local type int
./test.go:11:15: x.Bar undefined (type int has no field or method Bar)
```

Got (on master):
```console
$ go run ./cmd/yaegi/ test.go 
1
```

--- a/_test/type27.go
+++ b/_test/type27.go
@@ -1,0 +1,17 @@
+package main
+
+import "fmt"
+
+type Foo = int
+
+func (f Foo) Bar() int {
+	return f * f
+}
+
+func main() {
+	x := Foo(1)
+	fmt.Println(x.Bar())
+}
+
+// Error:
+// 7:1: cannot define new methods on non-local type int

--- a/_test/type28.go
+++ b/_test/type28.go
@@ -1,0 +1,17 @@
+package main
+
+import "fmt"
+
+type Foo = int
+
+func (f *Foo) Bar() int {
+	return *f * *f
+}
+
+func main() {
+	x := Foo(1)
+	fmt.Println(x.Bar())
+}
+
+// Error:
+// 7:1: cannot define new methods on non-local type int

--- a/_test/type29.go
+++ b/_test/type29.go
@@ -1,0 +1,12 @@
+package main
+
+var Foo int
+
+func (f Foo) Bar() int {
+	return 1
+}
+
+func main() {}
+
+// Error:
+// 5:1: cannot define new methods on non-local type int

--- a/_test/type30.go
+++ b/_test/type30.go
@@ -1,0 +1,12 @@
+package main
+
+var Foo *int
+
+func (f Foo) Bar() int {
+	return 1
+}
+
+func main() {}
+
+// Error:
+// 5:1: cannot define new methods on non-local type int

--- a/_test/type31.go
+++ b/_test/type31.go
@@ -14,4 +14,4 @@ func main() {
 }
 
 // Error:
-// 9:6: cannot define new methods on non-local type int
+// 5:1: cannot define new methods on non-local type int

--- a/_test/type31.go
+++ b/_test/type31.go
@@ -1,0 +1,17 @@
+package main
+
+import "fmt"
+
+func (f Foo) Bar() int {
+	return f * f
+}
+
+type Foo = int
+
+func main() {
+	x := Foo(1)
+	fmt.Println(x.Bar())
+}
+
+// Error:
+// 9:6: cannot define new methods on non-local type int

--- a/_test/type32.go
+++ b/_test/type32.go
@@ -14,4 +14,4 @@ func main() {
 }
 
 // Error:
-// 9:6: cannot define new methods on non-local type int
+// 5:1: cannot define new methods on non-local type int

--- a/_test/type32.go
+++ b/_test/type32.go
@@ -1,0 +1,17 @@
+package main
+
+import "fmt"
+
+func (f *Foo) Bar() int {
+	return *f * *f
+}
+
+type Foo = int
+
+func main() {
+	x := Foo(1)
+	fmt.Println(x.Bar())
+}
+
+// Error:
+// 9:6: cannot define new methods on non-local type int

--- a/_test/type33.go
+++ b/_test/type33.go
@@ -1,0 +1,11 @@
+package main
+
+func (f *Foo) Bar() int {
+	return *f * *f
+}
+
+func main() {
+}
+
+// Error:
+// 3:1: undefined: Foo

--- a/interp/ast.go
+++ b/interp/ast.go
@@ -93,7 +93,8 @@ const (
 	switchIfStmt
 	typeAssertExpr
 	typeDecl
-	typeSpec
+	typeSpec       // type A int
+	typeSpecAssign // type A = int
 	typeSwitch
 	unaryExpr
 	valueSpec
@@ -176,6 +177,7 @@ var kinds = [...]string{
 	typeAssertExpr:    "typeAssertExpr",
 	typeDecl:          "typeDecl",
 	typeSpec:          "typeSpec",
+	typeSpecAssign:    "typeSpecAssign",
 	typeSwitch:        "typeSwitch",
 	unaryExpr:         "unaryExpr",
 	valueSpec:         "valueSpec",
@@ -826,6 +828,10 @@ func (interp *Interpreter) ast(src, name string, inc bool) (string, *node, error
 			st.push(addChild(&root, anc, pos, typeAssertExpr, aTypeAssert), nod)
 
 		case *ast.TypeSpec:
+			if a.Assign.IsValid() {
+				st.push(addChild(&root, anc, pos, typeSpecAssign, aNop), nod)
+				break
+			}
 			st.push(addChild(&root, anc, pos, typeSpec, aNop), nod)
 
 		case *ast.TypeSwitchStmt:

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -149,26 +149,36 @@ func (interp *Interpreter) gta(root *node, rpath, importPath, pkgName string) ([
 				n.ident = ident
 				rcvr := n.child[0].child[0]
 				rtn := rcvr.lastChild()
-				typeName := rtn.ident
-				if typeName == "" {
-					// The receiver is a pointer, retrieve typeName from indirection
-					typeName = rtn.child[0].ident
-					elementType := sc.getType(typeName)
-					if elementType == nil {
+				typName := rtn.ident
+				if typName == "" {
+					// The receiver is a pointer, retrieve typName from indirection
+					typName = rtn.child[0].ident
+					sym, _, found := sc.lookup(typName)
+					if !found {
 						// Add type if necessary, so method can be registered
-						sc.sym[typeName] = &symbol{kind: typeSym, typ: &itype{name: typeName, path: pkgName, incomplete: true, node: rtn.child[0], scope: sc}}
-						elementType = sc.sym[typeName].typ
+						sym = &symbol{kind: typeSym, typ: &itype{name: typName, path: pkgName, incomplete: true, node: rtn.child[0], scope: sc}}
+						sc.sym[typName] = sym
 					}
+					if sym.kind != typeSym || (sym.node != nil && sym.node.kind == typeSpecAssign) {
+						err = n.cfgErrorf("cannot define new methods on non-local type %s", baseType(sym.typ).id())
+						return false
+					}
+					elementType := sym.typ
 					rcvrtype = ptrOf(elementType, withNode(rtn), withScope(sc))
 					rcvrtype.incomplete = elementType.incomplete
 					elementType.addMethod(n)
 				} else {
-					rcvrtype = sc.getType(typeName)
-					if rcvrtype == nil {
+					sym, _, found := sc.lookup(typName)
+					if !found {
 						// Add type if necessary, so method can be registered
-						sc.sym[typeName] = &symbol{kind: typeSym, typ: &itype{name: typeName, path: pkgName, incomplete: true, node: rtn, scope: sc}}
-						rcvrtype = sc.sym[typeName].typ
+						sym = &symbol{kind: typeSym, typ: &itype{name: typName, path: pkgName, incomplete: true, node: rtn, scope: sc}}
+						sc.sym[typName] = sym
 					}
+					if sym.kind != typeSym || (sym.node != nil && sym.node.kind == typeSpecAssign) {
+						err = n.cfgErrorf("cannot define new methods on non-local type %s", baseType(sym.typ).id())
+						return false
+					}
+					rcvrtype = sym.typ
 				}
 				rcvrtype.addMethod(n)
 				n.child[0].child[0].lastChild().typ = rcvrtype
@@ -265,7 +275,7 @@ func (interp *Interpreter) gta(root *node, rpath, importPath, pkgName string) ([
 				err = n.cfgErrorf("import %q error: %v", ipath, err)
 			}
 
-		case typeSpec:
+		case typeSpec, typeSpecAssign:
 			typeName := n.child[0].ident
 			var typ *itype
 			if typ, err = nodeType(interp, sc, n.child[1]); err != nil {
@@ -295,9 +305,13 @@ func (interp *Interpreter) gta(root *node, rpath, importPath, pkgName string) ([
 			}
 			sym, exists := sc.sym[typeName]
 			if !exists {
-				sc.sym[typeName] = &symbol{kind: typeSym}
+				sc.sym[typeName] = &symbol{kind: typeSym, node: n}
 			} else {
 				if sym.typ != nil && (len(sym.typ.method) > 0) {
+					if n.kind == typeSpecAssign {
+						err = n.cfgErrorf("cannot define new methods on non-local type %s", baseType(typ).id())
+						return false
+					}
 					// Type has already been seen as a receiver in a method function
 					for _, m := range sym.typ.method {
 						n.typ.addMethod(m)
@@ -305,7 +319,7 @@ func (interp *Interpreter) gta(root *node, rpath, importPath, pkgName string) ([
 				} else {
 					// TODO(mpl): figure out how to detect redeclarations without breaking type aliases.
 					// Allow redeclarations for now.
-					sc.sym[typeName] = &symbol{kind: typeSym}
+					sc.sym[typeName] = &symbol{kind: typeSym, node: n}
 				}
 			}
 			sc.sym[typeName].typ = n.typ
@@ -321,6 +335,17 @@ func (interp *Interpreter) gta(root *node, rpath, importPath, pkgName string) ([
 		sc.pop()
 	}
 	return revisit, err
+}
+
+func baseType(t *itype) *itype {
+	for {
+		switch t.cat {
+		case ptrT, aliasT:
+			t = t.val
+		default:
+			return t
+		}
+	}
 }
 
 // gtaRetry (re)applies gta until all global constants and types are defined.
@@ -346,7 +371,7 @@ func (interp *Interpreter) gtaRetry(nodes []*node, importPath, pkgName string) e
 	if len(revisit) > 0 {
 		n := revisit[0]
 		switch n.kind {
-		case typeSpec:
+		case typeSpec, typeSpecAssign:
 			if err := definedType(n.typ); err != nil {
 				return err
 			}

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -149,36 +149,26 @@ func (interp *Interpreter) gta(root *node, rpath, importPath, pkgName string) ([
 				n.ident = ident
 				rcvr := n.child[0].child[0]
 				rtn := rcvr.lastChild()
-				typName := rtn.ident
+				typName, typPtr := rtn.ident, false
 				if typName == "" {
-					// The receiver is a pointer, retrieve typName from indirection
-					typName = rtn.child[0].ident
-					sym, _, found := sc.lookup(typName)
-					if !found {
-						// Add type if necessary, so method can be registered
-						sym = &symbol{kind: typeSym, typ: &itype{name: typName, path: pkgName, incomplete: true, node: rtn.child[0], scope: sc}}
-						sc.sym[typName] = sym
-					}
-					if sym.kind != typeSym || (sym.node != nil && sym.node.kind == typeSpecAssign) {
-						err = n.cfgErrorf("cannot define new methods on non-local type %s", baseType(sym.typ).id())
-						return false
-					}
+					typName, typPtr = rtn.child[0].ident, true
+				}
+				sym, _, found := sc.lookup(typName)
+				if !found {
+					n.meta = n.cfgErrorf("undefined: %s", typName)
+					revisit = append(revisit, n)
+					return false
+				}
+				if sym.kind != typeSym || (sym.node != nil && sym.node.kind == typeSpecAssign) {
+					err = n.cfgErrorf("cannot define new methods on non-local type %s", baseType(sym.typ).id())
+					return false
+				}
+				rcvrtype = sym.typ
+				if typPtr {
 					elementType := sym.typ
 					rcvrtype = ptrOf(elementType, withNode(rtn), withScope(sc))
 					rcvrtype.incomplete = elementType.incomplete
 					elementType.addMethod(n)
-				} else {
-					sym, _, found := sc.lookup(typName)
-					if !found {
-						// Add type if necessary, so method can be registered
-						sym = &symbol{kind: typeSym, typ: &itype{name: typName, path: pkgName, incomplete: true, node: rtn, scope: sc}}
-						sc.sym[typName] = sym
-					}
-					if sym.kind != typeSym || (sym.node != nil && sym.node.kind == typeSpecAssign) {
-						err = n.cfgErrorf("cannot define new methods on non-local type %s", baseType(sym.typ).id())
-						return false
-					}
-					rcvrtype = sym.typ
 				}
 				rcvrtype.addMethod(n)
 				n.child[0].child[0].lastChild().typ = rcvrtype
@@ -375,7 +365,7 @@ func (interp *Interpreter) gtaRetry(nodes []*node, importPath, pkgName string) e
 			if err := definedType(n.typ); err != nil {
 				return err
 			}
-		case defineStmt:
+		case defineStmt, funcDecl:
 			if err, ok := n.meta.(error); ok {
 				return err
 			}

--- a/interp/interp_consistent_test.go
+++ b/interp/interp_consistent_test.go
@@ -107,7 +107,8 @@ func TestInterpConsistencyBuild(t *testing.T) {
 			file.Name() == "type29.go" || // expect error
 			file.Name() == "type30.go" || // expect error
 			file.Name() == "type31.go" || // expect error
-			file.Name() == "type32.go" { // expect error
+			file.Name() == "type32.go" || // expect error
+			file.Name() == "type33.go" { // expect error
 			continue
 		}
 

--- a/interp/interp_consistent_test.go
+++ b/interp/interp_consistent_test.go
@@ -101,7 +101,13 @@ func TestInterpConsistencyBuild(t *testing.T) {
 			file.Name() == "server.go" || // syntax parsing
 			file.Name() == "range9.go" || // expect error
 			file.Name() == "unsafe6.go" || // needs go.mod to be 1.17
-			file.Name() == "unsafe7.go" { // needs go.mod to be 1.17
+			file.Name() == "unsafe7.go" || // needs go.mod to be 1.17
+			file.Name() == "type27.go" || // expect error
+			file.Name() == "type28.go" || // expect error
+			file.Name() == "type29.go" || // expect error
+			file.Name() == "type30.go" || // expect error
+			file.Name() == "type31.go" || // expect error
+			file.Name() == "type32.go" { // expect error
 			continue
 		}
 

